src/rust/engine/fs/src/store.rs

```diff
@@ -30,39 +30,77 @@ impl Into<bazel_protos::remote_execution::Digest> for Digest {
 ///
 /// A content-addressed store of file contents, and Directories.
 ///
-/// Currently, Store only stores things locally on disk, but in the future it will gain the ability
-/// to fetch files from remote content-addressable storage too.
+/// Store keeps content on disk, and can optionally delegate to backfill its on-disk storage by
+/// fetching files from a remote server which implements the gRPC bytestream interface
+/// (see https://github.com/googleapis/googleapis/blob/master/google/bytestream/bytestream.proto)
+/// as specified by the gRPC remote execution interface (see
+/// https://github.com/googleapis/googleapis/blob/master/google/devtools/remoteexecution/v1test/)
+///
+/// In the future, it will gain the ability to write back to the gRPC server, too.
 ///
 #[derive(Clone)]
 pub struct Store {
   local: local::ByteStore,
+  remote: Option<remote::ByteStore>,
 }
 
 // Note that Store doesn't implement ByteStore because it operates at a higher level of abstraction,
 // considering Directories as a standalone concept, rather than a buffer of bytes.
 // This has the nice property that Directories can be trusted to be valid and canonical.
 // We may want to re-visit this if we end up wanting to handle local/remote/merged interchangably.
 impl Store {
-  pub fn new<P: AsRef<Path>>(path: P, pool: Arc<ResettablePool>) -> Result<Store, String> {
-    Ok(Store { local: local::ByteStore::new(path, pool)? })
+  ///
+  /// Make a store which only uses its local storage.
+  ///
+  pub fn local_only<P: AsRef<Path>>(path: P, pool: Arc<ResettablePool>) -> Result<Store, String> {
+    Ok(Store {
+      local: local::ByteStore::new(path, pool)?,
+      remote: None,
+    })
+  }
+
+  ///
+  /// Make a store which uses local storage, and if it is missing a value which it tries to load,
+  /// will attempt to back-fill its local storage from a remote CAS.
+  ///
+  pub fn backfills_from_remote<P: AsRef<Path>>(
+    path: P,
+    pool: Arc<ResettablePool>,
+    cas_address: String,
+  ) -> Result<Store, String> {
+    Ok(Store {
+      local: local::ByteStore::new(path, pool)?,
+      remote: Some(remote::ByteStore::new(cas_address)),
+    })
   }
 
   pub fn store_file_bytes(&self, bytes: Vec<u8>) -> BoxFuture<Digest, String> {
     self.local.store_bytes(EntryType::File, bytes)
   }
 
+  ///
+  /// Loads the bytes of the file with the passed fingerprint, and returns the result of applying f
+  /// to that value.
+  ///
   pub fn load_file_bytes_with<T: Send + 'static, F: Fn(&[u8]) -> T + Send + Sync + 'static>(
     &self,
     fingerprint: Fingerprint,
     f: F,
   ) -> BoxFuture<Option<T>, String> {
-    self.local.load_bytes_with(
+    let f_local = Arc::new(f);
```

Comment:
I'm not sure I see the reason for the creating the `Arc` in this function rather than in `load_bytes_with`. Could remove the generic parameter redundancy by creating the `Arc` there instead.

Reply:
load_bytes_with takes two different functions, one which is applied to local results, and one which is applied to remote results. For the file case, they are the same, but for the Directory case, they are different (so we only do the somewhat-expensive canonicalisation checks when we are about to store a Directory we got remotely, so can't trust).

So we can't Arc here, because the other caller doesn't pass in two copies of the same function.

What would make that more clear? :)

Also, if there's a cute way to clone or something so that I don't have to write out the function twice here, I'd love that...

Reply:
@illicitonion : Interesting. Certainly not a blocker... can fiddle with it later.